service/src/test/java/bio/terra/drshub/services/DrsApiFactoryTest.java

```diff
@@ -38,43 +45,44 @@ private DrsProvider createDrsProvider() {
 
   @Test
   void testMTlsConfigured() {
-    var mTlsConfig = MTlsConfig.create().setCertPath("fake.crt").setKeyPath("fake.key");
+    var mTlsConfig = MTlsConfig.create();
+    when(restTemplateFactory.makeMTlsRestTemplateWithPooling(mTlsConfig)).thenReturn(restTemplate);
 
-    drsApiFactory.getApiFromUriComponents(
-        URI_COMPONENTS, createDrsProvider().setMTlsConfig(mTlsConfig));
+    var drsApi =
+        drsApiFactory.getApiFromUriComponents(
+            URI_COMPONENTS, createDrsProvider().setMTlsConfig(mTlsConfig));
+    assertThat(drsApi.getApiClient(), is(apiClient));
 
-    verify(restTemplateFactory).makeMTlsRestTemplateWithPooling(any(MTlsConfig.class));
-    verify(restTemplateFactory, never()).makeRestTemplateWithPooling();
+    verify(drsApiClientFactory).createClient(restTemplate);
   }
 
   @Test
   void testMTlsNotConfigured() {
-    drsApiFactory.getApiFromUriComponents(URI_COMPONENTS, createDrsProvider());
+    when(restTemplateFactory.makeRestTemplateWithPooling()).thenReturn(restTemplate);
 
-    verify(restTemplateFactory).makeRestTemplateWithPooling();
-    verify(restTemplateFactory, never()).makeMTlsRestTemplateWithPooling(any(MTlsConfig.class));
+    var drsApi = drsApiFactory.getApiFromUriComponents(URI_COMPONENTS, createDrsProvider());
+    assertThat(drsApi.getApiClient(), is(apiClient));
+
+    verify(drsApiClientFactory).createClient(restTemplate);
   }
 
   @Test
   void testIndependentApiClients() {
     var drsProvider = createDrsProvider();
 
     // The first ApiClient created for a DrsProvider will populate the RestTemplate cache
-    var client1 = drsApiFactory.getApiFromUriComponents(URI_COMPONENTS, drsProvider);
+    when(restTemplateFactory.makeRestTemplateWithPooling()).thenReturn(restTemplate);
+    drsApiFactory.getApiFromUriComponents(URI_COMPONENTS, drsProvider);
     verify(restTemplateFactory).makeRestTemplateWithPooling();
-    verify(restTemplateFactory, never()).makeMTlsRestTemplateWithPooling(any(MTlsConfig.class));
 
     // Subsequent ApiClients created for the same DrsProvider will reuse the cached RestTemplate
-    var separateApiClient =
-        spy(drsApiFactory.getApiFromUriComponents(URI_COMPONENTS, drsProvider).getApiClient());
-    verifyNoMoreInteractions(restTemplateFactory);
-
-    // Setting the token or headers on an ApiClient will have no effect on ApiClients created for
-    // the same DrsProvider
-    client1.setBearerToken("bearer-token");
-    verifyNoInteractions(separateApiClient);
+    drsApiFactory.getApiFromUriComponents(URI_COMPONENTS, drsProvider);
+    // If there was a cache miss, this would fail as the API would be called twice.
+    verify(restTemplateFactory).makeRestTemplateWithPooling();
 
-    client1.setHeader("test-header", "test-value");
```

Comment:
These removed test assertions were providing real value, and were there in response to the bug where initial ApiClient caching inadvertently allowed headers to be shared between requests.  But in this new structure, this test is no longer the best place for them: we will want to test that DrsApiClientFactory creates new ApiClients each time, even for the same RestTemplate. I can take care of that in my main feature branch with this pulled in.

Reply:
I think the new test provides an equivalent test to the original. Instead of testing the behavior of calling `DrsApi.setHeader()` (which should be covered by a test on `DrsApi`), it verifies that the same rest template is being used for the first two clients, and a different one is being used for the last client, by looking at how often `DrsApiFactory` calls `makeRestTemplateWithPooling()`. It doesn't verify that a new client is made each time, but you could by asserting

```
    verify(drsApiClientFactory, times(3)).createClient(any(RestTemplate.class));
```

or by mocking `createClient()` to create a new client each time instead of returning the same object each time.